### DISCUSSION TITLE
Throwing from onConnect - what's the proper pattern?

### DISCUSSION THREAD

**Discussion by theogravity** [enisdenjo/graphql-ws#317] [Q&A] [answered]
If I throw from the `onConnect` handler, i'll get a message like this:

```
Internal error occurred during message handling. Please check your implementation
```

I get the sense that I shouldn't be throwing an error, but instead closing the socket with some kind of message. What's a good example of that?

Does this also apply to the other event listeners as well?

**Comment by theogravity** [comment]
The docs does say that I should be able to throw from it, and the socket will close appropriately, but should I be getting that specific error on the server-side? I'm throwing an `ApolloError` type, but I can throw a normal `Error` if required.

https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#onconnect

**Comment by enisdenjo** [comment, marked as answer]
Throwing from any of the server hooks close the WebSocket connection with a: `4500: Internal server error`. The close reason indicates that the server occurred an unexpected error, is terminal by nature and should not happen. Important thing to note is that the [client will report the `4500` close code immediately and not retry](https://github.com/enisdenjo/graphql-ws/blob/96058eb08f9c98ea4102f0aa90cc25f7be12f955/src/client.ts#L775).

`graphql-ws` doesn't have opinions on how closing connections on errors should look like, so it just offers a basic solution in case of thrown errors in unexpected places. [There are numerous close codes you can use on a WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/code), they're all in your hands.

With this liberty, you can catch an error in a hook, close with whatever code is agreed with a client for that specific error, have the client do something fancy; or just ignore the error all-together and proceed with execution, it's up to you. Note that `graphql-ws` handles cleanup when closing sockets abruptly, both on the server and client. Whenever you choose to `ctx.extra.socket.close(4321, 'My error')` in any of the hooks, the library will take care of ending active subscriptions and will perform proper disposal.

**Reply by theogravity** [reply]
Thank you!
